test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/StringTest.scala

```diff
@@ -255,13 +301,92 @@ class StringTest {
     assertThrows(classOf[IndexOutOfBoundsException], s.offsetByCodePoints(30, 2))
   }
 
+  @Test def substringBegin(): Unit = {
+    assertEquals("", "".substring(0))
+    assertEquals("", "foo".substring(3))
+    assertEquals("", "hello".substring(5))
+    assertEquals("lo", "hello".substring(3))
+    assertEquals("baz", "foo bar baz".substring(8))
+    assertEquals("foo bar baz", "foo bar baz".substring(0))
+  }
+
+  @Test def substringBeginIndexOutOfBounds(): Unit = {
+    assumeTrue("Assuming compliant StringIndexOutOfBounds",
+        hasCompliantStringIndexOutOfBounds)
+
+    assertThrows(classOf[StringIndexOutOfBoundsException],
+        "foo".substring(-1))
+    assertThrows(classOf[StringIndexOutOfBoundsException],
+        "foo".substring(4))
+    assertThrows(classOf[StringIndexOutOfBoundsException],
+        "foo".substring(15))
+
+    assertThrows(classOf[StringIndexOutOfBoundsException],
+        "".substring(-1))
+    assertThrows(classOf[StringIndexOutOfBoundsException],
+        "".substring(1))
+  }
+
+  @Test def substringBeginEnd(): Unit = {
+    assertEquals("", "".substring(0, 0))
+    assertEquals("", "foo".substring(3, 3))
+    assertEquals("", "hello".substring(3, 3))
+    assertEquals("lo", "hello".substring(3, 5))
+    assertEquals("bar", "foo bar baz".substring(4, 7))
+    assertEquals("foo bar baz", "foo bar baz".substring(0, 11))
```

Comment:
Also add a test case where the end index isn't the max?

Reply:
I assume you mean a case where the begin index is 0? Because otherwise the `"bar"` case already covers that.

Reply:
Ah, yes indeed. I did mean end index isn't max. But indeed that was already there :shrug: